### PR TITLE
[WIP] Python3 support - Django

### LINKED ISSUE #121: Python 3 support
We need to figure out how to get SymPy Live to support Python 3. I saw https://stackoverflow.com/questions/24229203/does-google-app-engine-support-python-3 but I couldn't figure out how to get it working. 

SymPy will drop Python 2 support at the end of 2019, so we won't be able to update SymPy Live after that unless we get Python 3 working.

### PR DESCRIPTION
**Pull request by prshnt19**
Fixes #121

### CODE REVIEW

**Comment by asmeurer** [comment]
What is cloud_sql_proxy?

**Comment by asmeurer** [comment]
What is cloud_sql_proxy?

**Comment by prshnt19** [comment]
Earlier AppEngine had its own database. But now we have to use Third-Party databases like MySQL hosted on Google Cloud. cloud_sql_proxy is used to connect to this Cloud MySQL shell from our device for development purposes.

**Comment by asmeurer** [comment]
I ask because it's a 13 MB binary file. Is it necessary to include it in the repo?

**Comment by prshnt19** [comment]
No, it's not necessary.

**Comment by asmeurer** [comment]
You'll need to squash the commit that removes it, so that it isn't included in the git history.

File: shell/views.py [hunk 1]
```diff
@@ -29,16 +29,30 @@
 definitions, and other values in the global and local namespaces can be used
 across commands.
 
-To use the shell in your app, copy shell.py, static/*, and templates/* into
+To use the shell in your app, copy shell.py, sympy/*, and templates/* into
```
Comment (prshnt19) [inline shell/views.py]: This is a mistake, I will correct it.

**Comment by prshnt19** [comment]
I have hosted it on AppEngine for testing. Works fine. Please have a look.
https://sympy-live-pr.appspot.com
ID : test
pass : test9876

**Comment by oscarbenjamin** [comment]
I notice that underscore for the last output doesn't work e.g.:
```
>>> x = 2
>>> x
2
>>> 3*_
6
```

**Comment by asmeurer** [comment]
The completion code seems to be broken. 

It may also be helpful if you push up a version of the docs to a github fork that points to this.

**Comment by prshnt19** [comment]
I have made changes to the code.
'_' working.
completion working

File: shell/migrations/0004_auto_20200225_1806.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    dependencies = [
+        ('shell', '0003_auto_20200209_1806'),
+    ]
+
+    operations = [
+        migrations.AlterModelOptions(
+            name='searches',
+            options={'verbose_name_plural': 'Searches'},
+        ),
+    ]
```
Comment (ivanistheone) [inline shell/migrations/0004_auto_20200225_1806.py]: When you're done with the dev on this PR, perhaps you could delete all these migrations and replace with a single one. It's inevitable that migrations will get messy over time, but at least initial one can be nice and simple ;)

**Comment by prshnt19** [comment]
> It may also be helpful if you push up a version of the docs to a github fork that points to this.

I didn't understand. Do I need to push docs for the changes I have made?

File: shell/migrations/0004_auto_20200225_1806.py (hunk 2 above)
Comment (prshnt19) [reply shell/migrations/0004_auto_20200225_1806.py]: Yeah Sure. Once done I will replace them with a single one.

**Comment by asmeurer** [comment]
The easiest way will be to build a version of the SymPy docs that point to this (you will need to update the Sphinx extension and edit the `doc/conf.py` in sympy to point to the updated version), then build the docs, create a repo on GitHub with a gh-pages branch, and push the built docs there (copy `_build/html` there after building the docs). Is that clear? I can provide more detailed instructions if you need them.

**Comment by prshnt19** [comment]
Thanks. Now it's clear. I will ask in case of any issue.

**Comment by prshnt19** [comment]
> It may also be helpful if you push up a version of the docs to a github fork that points to this.

https://prashant.rawat.tk/sympy-live-docs

**Comment by prshnt19** [comment]
ping @asmeurer

**Comment by oscarbenjamin** [comment]
If you go here:
https://prashant.rawat.tk/sympy-live-docs/tutorial/basic_operations.html
and run the second block you'll see:
```
>>> expr = cos(x) + 1
>>> expr.subs(x, y)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'sympy.core.expr' has no attribute 'subs'
```

**Comment by asmeurer** [comment]
I tried to verify the SymPy version with

```
import sympy
sympy.__version__
```

but I got the error `NameError: name 'sympy' is not defined`.

**Comment by prshnt19** [comment]
https://sympy-live-pr.appspot.com does not support anonymous sessions for now. It needs a login. Anonymous sessions will resolve both issues. I am working on it.

**Comment by prshnt19** [comment]
Anonymous User Issue fixed. You may test.

**Comment by sudoWin** [comment]
This doesn't work

```python
In [3]: A = Matrix([[12, -51, 4], [6, 167, -68], [-4, 24, -41]])                                                                                                                                            

In [4]: Q, R = A.QRdecomposition()
```

**Comment by prshnt19** [comment]
Sympy has predefined `Q`. when we execute `from sympy import *`, `Q` is also imported. This error can be prevented by using `q` instead or reinitialising `Q`, like:
```
Q=1      # or anything else 
A = Matrix([[12, -51, 4], [6, 167, -68], [-4, 24, -41]])
Q, R = A.QRdecomposition()```

**Comment by sudoWin** [comment]
Ohh, So probably the example in the docs which has "Run Code Block in Sympy Live" also needs to be changed.

**Comment by prshnt19** [comment]
Thanks for pointing this out.

File: shell/migrations/0004_auto_20200225_1806.py (hunk 2 above)
Comment (prshnt19) [reply shell/migrations/0004_auto_20200225_1806.py]: I have deleted all the migrations. Replaced them with a single migration.

**Comment by prshnt19** [comment]
ping @asmeurer @oscarbenjamin

**Comment by asmeurer** [comment]
Sorry I haven't had a chance to review this. I will try to do so soon. 

Meanwhile you shouldn't let the lack of review here slow you down on any potential GSoC proposal.

File: index.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  properties:
+  - name: private
+  - name: timestamp
+    direction: desc
+
+- kind: Searches
+  properties:
+  - name: user_id
+  - name: timestamp
+    direction: desc
```
Comment (aktech) [inline index.yaml:23]: What is the purpose of these? I highly doubt that it is supported in Python3 runtime.

File: README.rst [hunk 4]
```diff
@@ -48,6 +48,82 @@ We use submodules to include external libraries in sympy-live::
...
+    $ sudo apt-get install python3-dev
+    $ sudo systemctl start mysql
+    $ sudo mysql
+
+This will open mysql. You will now have to create a new database with name 
+{database_name}. Create a new user {user_name} and password {user_password}.
+Grant all previleges to {user_name} on {database_name}::
+
+    mysql> CREATE DATABASE {database_name};
+    mysql> UPDATE mysql.user SET authentication_string = PASSWORD('{user_password}') WHERE User = '{user_name}';
```
Comment (aktech) [inline README.rst:85]: This doesn't works for me, 
```
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ...
```
I think this syntax varies with mysql version.

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Django==3.0.4
+django-cors-headers==3.2.1
+gunicorn==20.0.4
+matplotlib==3.2.0
+mysqlclient==1.4.6
+numpy==1.18.1
+scipy==1.4.1
```
Comment (aktech) [inline requirements.txt:7]: Where do we use these?

File: index.yaml (hunk 3 above)
Comment (prshnt19) [reply index.yaml:23]: It was automatically generated while running dev_appserver for Python2 local testing. I think it should be removed.

**Comment by prshnt19** [comment]
This PR is a result of my attempt toward updating the Python2 to Python3 on Google AppEngine. AppEngine features changes with the selection of a language. For Python2 AppEngine uses its own libraries for DataBase, Users, etc. But while deploying with Python3 we have to use Third Party Libraries only. So to upgrade from Python2 to Python3 we also have to make a decision for Framework and DataBase of our own. This leaves us at implementing these task:
* Upgrading Python2 -> Python3
* Deciding Web framework
* Adding a SQL database

So I went with Django and SQL. I used Cloud SQL for testing purpose because it was easy to set up, with no aim of using it in the long run. I am open to any alternatives. I haven't used BigQuery earlier but I do have some doubts about it. BigQuery is useful in cases where tables are large enough and row modification is not there. But our sessions will be storing data which will be modified with every statement we execute. So, in that case we might not be able to use BigQuery. Please correct me if I am wrong at some point.

I think updating the sessions with the pickled data is a bit easier in Django. Django gives many options to modify sessions with ease. With also being used in SymPy Gamma, I think Django is a better alternative to Flask. I also want to know the framework you think will be best suitable. The work on this PR is still in progress and a lot of things need to be figured out. Thanks again for your feedback.

File: README.rst (hunk 4 above)
Comment (prshnt19) [reply README.rst]: I have updated the README. If it still doesn't work out for you please check MySQL documentation.

File: requirements.txt (hunk 5 above)
Comment (prshnt19) [reply requirements.txt:7]: They were to be used in case we make SymPy Live more generic. I think I should remove them and address that issue separately.

**Comment by aktech** [comment]
> This PR is a result of my attempt toward updating the Python2 to Python3 on Google AppEngine. AppEngine features changes with the selection of a language. For Python2 AppEngine uses its own libraries for DataBase, Users, etc. But while deploying with Python3 we have to use Third Party Libraries only. So to upgrade from Python2 to Python3 we also have to make a decision for Framework and DataBase of our own. 

As I mentioned above, this PR is trying to solve multiple problems at once, the above doesn't needs to happen in one PR and it shouldn't as it will be hard to make right decisions and review, we need to do it one step at a time, by that I mean we need to prepare the app for it to migrate to Python 3.

So, for example currently this app uses bundled app engine services like datastore and uses Python2 NDB Client library to connect to it, so the first thing  that needs upgrading is updating this to use [Cloud NDB](https://cloud.google.com/appengine/docs/standard/python3/migrating-to-cloud-ndb), which works on Python 2 as well, so we don't need to upgrade Python for this step. Similarly you can check out if there is any other bundled services it is using , which needs replacing and so on.

Here is a step by step guide:
https://cloud.google.com/appengine/docs/standard/python/migrate-to-python3/

> So I went with Django and SQL. I used Cloud SQL for testing purpose because it was easy to set up, with no aim of using it in the long run.

Choosing framework would come up in later steps.

>  BigQuery is useful in cases where tables are large enough and row modification is not there. But our sessions will be storing data which will be modified with every statement we execute. 

I don't mean to store things like sessions, which we constantly need to query in bigquery, its not great for reading in real time systems. For sessions we can still use Datastore, which we already use at the moment. If that's a problem or not it's a question for later when we have done the upgrade. Bigquery could be useful for storing search queries, the things we don't need to lookup, although I would still use Datastore for the same to start with and if it grows then moving to Bigquery later.

> I think updating the sessions with the pickled data is a bit easier in Django. Django gives many options to modify sessions with ease. With also being used in SymPy Gamma, I think Django is a better alternative to Flask. I also want to know the framework you think will be best suitable. The work on this PR is still in progress and a lot of things need to be figured out. Thanks again for your feedback.

We will be more able to answer the choice of framework, when we have completed the steps preceding it, as it will be required when we move to Python3, which is a few steps away. My initial hunch is it won't matter much as Datastore is NoSQL database and Django doesn't supports  NoSQL Database natively, but don't worry about it at the moment. (As a side note, we're discussing if we can afford Cloud SQL or not, as soon as we know, I will let you know.)

**Comment by aktech** [comment]
> As a side note, we're discussing if we can afford Cloud SQL or not, as soon as we know, I will let you know.

To let you know, as per discussions with Aaron and Certik we have decided that we're not getting Cloud SQL as it doesn't justifies the cost at the moment.

**Comment by prshnt19** [comment]
OK, I got your point for step by step updating. I will start working on updating to Cloud NDB first. While working on it We can figure out other bundles which need to be updated. Is there anything else which I  might need to know before starting updating Cloud NDB? Or should I start working on it?

**Comment by aktech** [comment]
>OK, I got your point for step by step updating. I will start working on updating to Cloud NDB first. While working on it We can figure out other bundles which need to be updated.

This link has all the information regarding incremental upgrade:
https://cloud.google.com/appengine/docs/standard/python/migrate-to-python3/

>  Is there anything else which I might need to know before starting updating Cloud NDB? Or should I start working on it?

Apart from what's mentioned in the above, link I cannot think of anything else. I worked on the Cloud NDB migration recently in Gamma, the following PR should help you get an idea of what needs to be done:

https://github.com/sympy/sympy_gamma/pull/140

**Pull request closed without merging**